generator/camera_info_collector.cpp

```diff
@@ -0,0 +1,340 @@
+#include "generator/camera_info_collector.hpp"
+
+#include "coding/pointd_to_pointu.hpp"
+#include "coding/reader.hpp"
+#include "coding/varint.hpp"
+
+#include "base/checked_cast.hpp"
+
+#include <algorithm>
+
+namespace generator
+{
+// TODO (@gmoryes) add "inline static ..." after moving to c++17
+uint32_t constexpr CamerasInfoCollector::kLatestVersion;
+double constexpr CamerasInfoCollector::Camera::kEqualityEps;
+double constexpr CamerasInfoCollector::kMaxDistFromCameraToClosestSegmentMeters;
+double constexpr CamerasInfoCollector::kSearchCameraRadiusMeters;
+uint32_t constexpr CamerasInfoCollector::kMaxCameraSpeedKmpH;
+
+CamerasInfoCollector::CamerasInfoCollector(std::string const & dataFilePath,
+                                           std::string const & camerasInfoPath,
+                                           std::string const & osmIdsToFeatureIdsPath)
+{
+  std::map<osm::Id, uint32_t> osmIdToFeatureId;
+  if (!routing::ParseOsmIdToFeatureIdMapping(osmIdsToFeatureIdsPath, osmIdToFeatureId))
+  {
+    LOG(LWARNING, ("An error happened while parsing feature id to osm ids mapping from file:",
+                   osmIdsToFeatureIdsPath));
+
+    m_valid = false;
+    return;
+  }
+
+  if (!ParseIntermediateInfo(camerasInfoPath, osmIdToFeatureId))
+  {
+    LOG(LWARNING, ("Unable to parse intermediate file(", camerasInfoPath, ") about cameras info"));
+    m_valid = false;
+    return;
+  }
+
+  platform::LocalCountryFile file = platform::LocalCountryFile::MakeTemporary(dataFilePath);
+  FrozenDataSource dataSource;
+  auto registerResult = dataSource.RegisterMap(file);
+  if (registerResult.second != MwmSet::RegResult::Success)
+  {
+    LOG(LWARNING, ("Unable to RegisterMap:", dataFilePath));
+    m_valid = false;
+    return;
+  }
+
+  for (auto & camera : m_cameras)
+  {
+    camera.ParseDirection();
+    camera.FindClosestSegment(dataSource, registerResult.first);
+  }
+}
+
+void CamerasInfoCollector::Serialize(FileWriter & writer) const
+{
+  // Some cameras has several ways related to them, each way has featureId and segmentId.
+  // We should from each camera such type, with for example 3 ways, do 3 cameras with 1 way.
```

Comment:
Please rephrase this sentence.

Reply:
`We should from each camera such type with N ways do N cameras with 1 way.`
ok?

Reply:
We create N cameras with one way attached from a camera that has N ways attached.

Reply:
s/has/have/

The fact that each way has a corresponding featureId and segmentId does not seem relevant here.

Reply:
done